includes/class-wcpdf-endpoint.php

```diff
@@ -0,0 +1,107 @@
+<?php
+namespace WPO\WC\PDF_Invoices;
+
+use WPO\WC\PDF_Invoices\Compatibility\Order as WCX_Order;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+if ( ! class_exists( '\\WPO\\WC\\PDF_Invoices\\Endpoint' ) ) :
+
+class Endpoint {
+
+	public $action = 'generate_wpo_wcpdf';
+
+	public function __construct() {
+		if ( $this->is_enabled() ) {
+			add_action( 'init', array( $this, 'add_endpoint' ) );
+			add_action( 'query_vars', array( $this, 'add_query_vars' ) );
+			add_action( 'parse_request', array( $this, 'handle_document_requests' ) );
+		}
+	}
+
+	public function is_enabled() {
+		$debug_settings = get_option( 'wpo_wcpdf_settings_debug', array() );
+
+		if ( isset( $debug_settings['pretty_document_links'] ) ) {
+			return true;
+		} else {
+			return false;
+		}
+	}
+
+	public function get_identifier() {
+		return apply_filters( 'wpo_wcpdf_pretty_document_link_identifier', 'wcpdf' );
+	}
+	
+	public function add_endpoint() {
+		add_rewrite_rule(
+			'^'.$this->get_identifier().'/([^/]*)/([^/]*)?',
+			'index.php?action=generate_wpo_wcpdf&document_type=$matches[1]&order_ids=$matches[2]',
+			'top'
+		);
+	}
+
+	public function add_query_vars( $vars ) {
+		$vars[] = 'action';
+		$vars[] = 'document_type';
+		$vars[] = 'order_ids';
+		return $vars;
+	}
+
+	public function handle_document_requests() {
+		global $wp;
+
+		if ( ! empty( $wp->query_vars['action'] ) && strpos( $this->action, $wp->query_vars['action'] ) !== false ) {
+			if ( ! empty( $wp->query_vars['document_type'] ) && ! empty( $wp->query_vars['order_ids'] ) ) {
+				$_REQUEST['action']        = $this->action;
+				$_REQUEST['document_type'] = sanitize_text_field( $wp->query_vars['document_type'] );
+				$_REQUEST['order_ids']     = sanitize_text_field( $wp->query_vars['order_ids'] );
+				
+				do_action( 'wp_ajax_' . $this->action );
+			}
+		}
+	}
+
+	public function get_document_link( $order, $document_type, $additional_vars = array() ) {
+		if ( empty( $order ) || empty( $document_type ) ) {
+			return '';
+		}
+
+		if ( $this->is_enabled() ) {
+			$parameters = array(
+				$this->get_identifier(),
+				$document_type,
+				WCX_Order::get_id( $order ),
+			);
+			$document_link = trailingslashit( get_home_url() ) . implode( '/', $parameters );
+		} else {
+			$document_link = add_query_arg( array(
+				'action'        => $this->action,
+				'document_type' => $document_type,
+				'order_ids'     => WCX_Order::get_id( $order ),
+			), admin_url( 'admin-ajax.php' ) );
+		}
+
+		// handle guest access
+		$debug_settings = get_option( 'wpo_wcpdf_settings_debug', array() );
+		if ( is_user_logged_in() ) {
+			$document_link = add_query_arg( '_wpnonce', wp_create_nonce( $this->action ), $document_link );
+		} elseif ( ! is_user_logged_in() && isset( $debug_settings['guest_access'] ) ) {
+			$document_link = add_query_arg( 'order_key', $order->get_order_key(), $document_link );
+		}
```

Comment:
Can you not handle this in the permalink structure? I think the pretty URL should not have query args for the basic parameters that are always required. Right now you have
```
{$identifier}/{$document_type}/{$order_id}?_wpnonce=xxxx
```
but I think you could just add a 4th parameter that acts as a universal access key - checking first whether it matches the nonce, and if not, then check the `order_key`:
```
{$identifier}/{$document_type}/{$order_id}/{$access_key}
```